internal/orchestrator/reorg_handler_test.go

```diff
@@ -0,0 +1,502 @@
+		{Number: big.NewInt(104), Hash: "hash104", ParentHash: "hash103"},
+		{Number: big.NewInt(103), Hash: "hash103", ParentHash: "hash102"},
+		{Number: big.NewInt(102), Hash: "hash102", ParentHash: "hash101"},
+		{Number: big.NewInt(101), Hash: "hash101", ParentHash: "hash100"},
+		{Number: big.NewInt(100), Hash: "hash100", ParentHash: "hash99"},
+	}, nil)
+
+	mockRPC.EXPECT().GetBlocks([]*big.Int{big.NewInt(106), big.NewInt(105), big.NewInt(104), big.NewInt(103), big.NewInt(102), big.NewInt(101), big.NewInt(100)}).Return([]rpc.GetBlocksResult{
+		{BlockNumber: big.NewInt(106), Data: common.Block{Hash: "hash106", ParentHash: "hash105"}},
+		{BlockNumber: big.NewInt(105), Data: common.Block{Hash: "hash105", ParentHash: "hash104"}},
+		{BlockNumber: big.NewInt(104), Data: common.Block{Hash: "hash104", ParentHash: "hash103"}},
+		{BlockNumber: big.NewInt(103), Data: common.Block{Hash: "hash103", ParentHash: "hash102"}},
+		{BlockNumber: big.NewInt(102), Data: common.Block{Hash: "hash102", ParentHash: "hash101"}},
+		{BlockNumber: big.NewInt(101), Data: common.Block{Hash: "hash101", ParentHash: "hash100"}},
+		{BlockNumber: big.NewInt(100), Data: common.Block{Hash: "hash100", ParentHash: "hash99"}},
+	})
+
+	mockRPC.EXPECT().GetFullBlocks([]*big.Int{big.NewInt(105), big.NewInt(106)}).Return([]rpc.GetFullBlockResult{
+		{BlockNumber: big.NewInt(105), Data: common.BlockData{}},
+		{BlockNumber: big.NewInt(106), Data: common.BlockData{}},
+	})
+
+	mockMainStorage.EXPECT().DeleteBlockData(big.NewInt(1), mock.MatchedBy(func(blocks []*big.Int) bool {
+		return len(blocks) == 2
+	})).Return(nil)
+	mockMainStorage.EXPECT().InsertBlockData(mock.MatchedBy(func(data *[]common.BlockData) bool {
+		return data != nil && len(*data) == 2
+	})).Return(nil)
+
+	handler := NewReorgHandler(mockRPC, mockStorage)
+	mostRecentBlockChecked, err := handler.RunFromBlock(big.NewInt(109))
+
+	assert.NoError(t, err)
+	assert.Equal(t, big.NewInt(109), mostRecentBlockChecked)
+}
+
+func TestHandleReorgWithLatestBlockReorged(t *testing.T) {
+	defer func() { config.Cfg = config.Config{} }()
+	config.Cfg.ReorgHandler.BlocksPerScan = 10
+
+	mockRPC := mocks.NewMockIRPCClient(t)
+	mockMainStorage := mocks.NewMockIMainStorage(t)
+	mockOrchestratorStorage := mocks.NewMockIOrchestratorStorage(t)
+
+	mockStorage := storage.IStorage{
+		MainStorage:         mockMainStorage,
+		OrchestratorStorage: mockOrchestratorStorage,
+	}
+
+	mockRPC.EXPECT().GetChainID().Return(big.NewInt(1))
+	mockRPC.EXPECT().GetBlocksPerRequest().Return(rpc.BlocksPerRequestConfig{Blocks: 100})
+	mockOrchestratorStorage.EXPECT().GetLastReorgCheckedBlockNumber(big.NewInt(1)).Return(big.NewInt(100), nil)
+
+	mockMainStorage.EXPECT().LookbackBlockHeaders(big.NewInt(1), 10, big.NewInt(109)).Return([]common.BlockHeader{
+		{Number: big.NewInt(109), Hash: "hash109", ParentHash: "hash108a"}, // <-- fork starts here
+		{Number: big.NewInt(108), Hash: "hash108", ParentHash: "hash107"},
```

Comment:
didn't really understand this one, where is block with `hash108` then?

Reply:
ah, `hash108a` is an old one and thus block referring to this parent will be replaced, right?

Reply:
Yes, basically we detect a mismatch between block 108 and 109 (its parent references `108a`). So we poll the blocks 100 - 109 again to get the latest state about them.